### PR TITLE
Track exec sessions and relax exec locking

### PR DESCRIPTION
**Pull request by mheon**
Obsoletes #394 

Almost a wholescale rewrite of our Exec() code with the goal of reducing the locking around it, so that other commands can use a container while Exec() is running (e.g. `podman ps` will not hang if `podman exec` is running on a container).

### CODE REVIEW

File: libpod/container.go [hunk 1]
```diff
@@ -588,6 +599,50 @@ func (c *Container) PID() (int, error) {
...
+	if !c.locked {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+
+		if err := c.syncContainer(); err != nil {
+			return nil, err
+		}
+	}
+
+	ids := make([]string, 0, len(c.state.ExecSessions))
```
Comment (rhatdan) [inline libpod/container.go]: Why do you do this rather then just return the ExecSessions? Is this basically creating a snapshot of the execsessions?

File: libpod/container_api.go [hunk 2]
```diff
@@ -227,27 +237,111 @@ func (c *Container) Exec(tty, privileged bool, env, cmd []string, user string) e
...
+	}
+
+	if err := execCmd.Start(); err != nil {
+		return errors.Wrapf(err, "error starting exec command for container %s", c.ID())
+	}
+
+	pidFile := c.execPidPath(sessionID)
+	const pidWaitTimeout = 250
+
+	// Wait until runc makes the pidfile
```
Comment (rhatdan) [inline libpod/container_api.go]: NIT: sed s/runc/oci runitme/g

**Comment by rhatdan** [comment]
@mtrmac @runcom @baude PTAL

File: libpod/container.go (hunk 1 above)
Comment (mheon) [reply libpod/container.go]: I think most people are just going to be interested in whether there are active sessions (which are causing the container to not be able to be removed without -f), not the specifics of what's running in the container.

**Comment by baude** [comment]
did this fix issue #402 and if so, how about an integration test?

**Comment by mheon** [comment]
@baude Checking, but I suspect it was not a complete fix

**Comment by mheon** [comment]
So we are improved from our current state, in that we actually report errors. But we are returning the wrong error codes - we print out "Error -1" and return -127 for example

@baude Should we fix that here or merge this and fix separately?

**Comment by rhatdan** [comment]
@mheon lets do it separately.

**Comment by baude** [comment]
i only brought it up because I had Ed's issue in the back of my mind.  different PR is fine.

**Comment by rhatdan** [comment]
@baude are you ok with this patch?

**Comment by rhatdan** [comment]
I would like to get this merged so we can get it into tomorrows podman release.

**Comment by mheon** [comment]
Agree. I'll pester @baude

**Comment by baude** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**